swap/htlc_test.go

```diff
@@ -306,7 +323,291 @@ func TestHtlcV2(t *testing.T) {
+		PkScript: []byte{
+			0, 20, 2, 141, 221, 230, 144,
+			171, 89, 230, 219, 198, 90, 157,
+			110, 89, 89, 67, 128, 16, 150, 186,
+		},
+		Value: value,
+	}}
+
+	prevOutFetcher := txscript.NewCannedPrevOutputFetcher(
+		p2trPkScript, value,
+	)
+	hashCache := txscript.NewTxSigHashes(
+		tx, prevOutFetcher,
+	)
+
+	signTx := func(tx *wire.MsgTx, privateKey *secp.PrivateKey,
+		leaf txscript.TapLeaf) []byte {
+
+		sig, err := txscript.RawTxInTapscriptSignature(
+			tx, hashCache, 0, value, p2trPkScript, leaf,
+			txscript.SigHashDefault, privateKey,
+		)
+		require.NoError(t, err)
+
+		return sig
+	}
+
+	testCases := []struct {
+		name    string
+		witness func(*testing.T) wire.TxWitness
+		valid   bool
+	}{
+		{
+			// Receiver can spend with valid preimage.
+			"success case spend with valid preimage",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = htlc.SuccessSequence()
+				tx.LockTime = uint32(cltvExpiry)
+
+				trHtlc, ok := htlc.HtlcScript.(*HtlcScriptV3)
+				require.True(t, ok)
+
+				sig := signTx(
+					tx, receiverPrivKey,
+					txscript.NewBaseTapLeaf(
+						trHtlc.SuccessScript,
+					),
+				)
+				witness, err := htlc.genSuccessWitness(
+					sig, preimage,
+				)
+				require.NoError(t, err)
+
+				return witness
+			}, true,
+		},
+		{
+			// Receiver can't spend with the valid preimage and with
+			// zero sequence.
+			"success case no spend with valid preimage and zero sequence",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = 0
+
+				trHtlc, ok := htlc.HtlcScript.(*HtlcScriptV3)
+				require.True(t, ok)
+
+				sig := signTx(
+					tx, receiverPrivKey,
+					txscript.NewBaseTapLeaf(
+						trHtlc.SuccessScript,
+					),
+				)
+				witness, err := htlc.genSuccessWitness(
+					sig, preimage,
+				)
+				require.NoError(t, err)
+
+				return witness
+			}, false,
+		},
+		{
+			// Sender can't spend when haven't yet timed out.
+			"timeout case no spend before timeout",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = htlc.SuccessSequence()
+				tx.LockTime = uint32(cltvExpiry) - 1
+
+				trHtlc, ok := htlc.HtlcScript.(*HtlcScriptV3)
+				require.True(t, ok)
+
+				sig := signTx(
+					tx, senderPrivKey,
+					txscript.NewBaseTapLeaf(
+						trHtlc.TimeoutScript,
+					),
+				)
+
+				witness, err := htlc.GenTimeoutWitness(sig)
+				require.NoError(t, err)
+
+				return witness
+			}, false,
+		},
+		{
+			// Sender can spend after timeout.
+			"timeout case spend after timeout",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = htlc.SuccessSequence()
+				tx.LockTime = uint32(cltvExpiry)
+
+				trHtlc, ok := htlc.HtlcScript.(*HtlcScriptV3)
+				require.True(t, ok)
+
+				sig := signTx(
+					tx, senderPrivKey,
+					txscript.NewBaseTapLeaf(
+						trHtlc.TimeoutScript,
+					),
+				)
+
+				witness, err := htlc.GenTimeoutWitness(sig)
+				require.NoError(t, err)
+
+				return witness
+			}, true,
+		},
+		{
+			// Receiver can't spend after timeout.
```

Comment:
nit: this comment is slightly wrong (in the htlc v2 tests too) since the receiver can spend after timeout with the preimage, just can't spend using the timeout path.